crates/ty_python_semantic/src/semantic_index/builder.rs

```diff
@@ -1453,6 +1449,43 @@ impl<'ast> Visitor<'ast> for SemanticIndexBuilder<'_, 'ast> {
                 self.current_use_def_map_mut()
                     .record_node_reachability(NodeKey::from_node(node));
 
+                // If we see:
+                //
+                // * `from .x.y import z` (must be relative!)
+                // * And we are in an `__init__.py(i)` (hereafter `thispackage`)
+                // * And this is the first time we've seen `from .x` in this module
+                //
+                // We introduce a local definition `x = <module 'thispackage.x'>` that occurs
+                // before the `z = ...` declaration the import introduces. This models the fact
+                // that the *first* time that you import 'thispackage.x' the python runtime creates
+                // `x` as a variable in the global scope of `thispackage`.
+                //
+                // This is not a perfect simulation of actual runtime behaviour for *various*
+                // reasons but it works well for most practical purposes. In particular it's nice
+                // that `x` can be freely overwritten, and that we don't assume that an import
+                // in one function is visible in another function.
+                //
+                // TODO: Also support `from thispackage.x.y import z`?
+                // TODO: `seen_submodule_imports` should be per-scope and not per-file
+                // (if two functions import `.x`, they both should believe `x` is defined)
+                if node.level == 1
+                    && let Some(submodule) = &node.module
+                    && let Some(parsed_submodule) = ModuleName::new(submodule.as_str())
+                    && let Some(direct_submodule) = parsed_submodule.components().next()
+                    && self.file.is_package(self.db)
+                    && !self.seen_submodule_imports.contains(direct_submodule)
+                {
+                    self.seen_submodule_imports
+                        .insert(direct_submodule.to_owned());
+
+                    let direct_submodule_name = Name::new(direct_submodule);
+                    let symbol = self.add_symbol(direct_submodule_name);
+                    self.add_definition(
+                        symbol.into(),
+                        ImportFromImplicitDefinitionNodeRef { node, submodule },
```

Comment:
I don't know that it matters, since the from-import node itself is otherwise unused as a definition NodeKey, and there can only be one implicit definition per from-import, but I'm curious whether we could attach this definition to the Identifier for the submodule itself rather than to the entire import-from node? It seems slightly more intuitive/clearer. Maybe we can't do this because `Identifier` isn't a real AST node? (I don't recall...)